torch/csrc/api/include/torch/nn/module.h

```diff
@@ -0,0 +1,130 @@
+#pragma once
+
+#include <torch/tensor.h>
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace at {
+enum class ScalarType;
+} // namespace at
+
+namespace torch { namespace nn {
+
+struct Archive;
+struct BufferCursor;
+struct ModuleCursor;
+struct ParameterCursor;
+
+/// The base class for all torch modules.
+class Module {
+ public:
+  /// Tells the base `Module` about the name of the submodule.
+  explicit Module(std::string name);
```

Comment:
Do we ever expect `name` to change at run time? If I understand correctly this is a replacement for the class name in Python, so it shouldn't change (and we could take a `const char*`)

Reply:
Wouldn't it be more convenient to specify a name when adding a module to another one, i.e. similar to the Python API's `add_module()` function? As far as I can see hat's the only place where you actually need a unique name.

Reply:
I think the class name might be useful for printing?

Reply:
Yes it was primarily thought for pretty-printing, this constructor only gets invoked from the derived class constructor, to provide it with a name

Reply:
But it's also used for the maps in the module so you'll have to make sure it's unique, e.g. with the code right now you can't add two LSTMs to a module.